File: c/prime.c
#include <stdio.h>
#include <stdbool.h>

float d_sqrt(float n) {
    int i;
    float x, y;
    x = n * 0.5;
    y = n;
    i = *(int *) &y;
    i = 0x5f3759df - (i >> 1);
    y = *(float *) &i;
    y = y * (1.5 - (x * y * y));
    y = y * (1.5 - (x * y * y));
    return n * y;
}

int countPrimes(int n) {

    int count = 1;
    int i = 3;

    for (; i <= n; i += 2) {
        
        bool is_prime = true;
        float sqrti = d_sqrt(i);
	int t = 3;

        while (t <= sqrti) {
            if (i % t == 0) {
                is_prime = false;
                break;
            }
            t += 2;
        }
        
        if (is_prime == true) {
            count++;
        }
    }

    return count;

}

int main(int args, char **argv) {
    printf("There are %d prime numbers between 0 and 10000000\n", countPrimes(10000000));

    return 0;
}

File: nim/prime.nim
import math

proc countPrimes(max: int): int =
    var n = 1
    var i = 3

    while i < max+1:
        var isPrime = true
        let sqrti = int(sqrt(float(i)))
        var j = 3

        while j <= sqrti:
            if i mod j == 0:
                isPrime = false
                break
            j = j+2

        if isPrime == true:
            n = n+1

        i = i+2

    result = n

echo countPrimes(10_000_000)
